students/johnn/session09/mailroom.py

```diff
@@ -0,0 +1,688 @@
+
+    def add_donation(self, value):
+        """ add a single donation """
+        today = datetime.datetime.utcnow().date().isoformat() + "Z"
+        try:
+            value = float(value)
+        except ValueError:
+            raise ValueError("Donations must be values.")
+        self._donations.append( { "amount": value, "date": today } )
+
+    @property
+    def number_donations(self):
+        """ return the number of donations """
+        return len(self._donations)
+
+    @property
+    def total_donations(self):
+        """ return a sum of the donations """
+        total_donations = 0
+        for donation in self.donations:
+            total_donations += donation["amount"]
+        return round(total_donations, 2)
+
+    @property
+    def average_donations(self):
+        """ return the average donation amount """
+        try:
+            return round(self.total_donations / self.number_donations, 2)
+        except ZeroDivisionError:
+            return 0
+
+    @property
+    def first_name(self):
+        """ return the first name """
+        return self._first_name
+
+    @first_name.setter
+    def first_name(self, value):
+        """ set the first name """
+        self._first_name = value.title()
+
+    @property
+    def middle_name(self):
+        """ return the middle name """
+        return self._middle_name
+
+    @middle_name.setter
+    def middle_name(self, value):
+        """ set the middle name """
+        self._middle_name = value.title()
+
+    @property
+    def last_name(self):
+        """ return the last name """
+        return self._last_name
+
+    @last_name.setter
+    def last_name(self, value):
+        """ set the last name """
+        self._last_name = value.title()
+
+    @property
+    def suffix(self):
+        """ return the suffix """
+        return self._suffix
+
+    @suffix.setter
+    def suffix(self, value):
+        """ set the suffix """
+        # special case roman numbers to all caps
+        if value.lower() in ["i", "ii", "iii", "iv", "v"]:
+            self._suffix=value.upper()
+        else:
+            self._suffix=value.title()
+
+    @property
+    def informal_name(self):
+        """ return full name minus the suffix """
+        return " ".join(filter(None, [self.first_name, self.middle_name, self.last_name]))
+
+    @property
+    def full_name(self):
+        """ return the formal, full name """
+        # if suffix, add with a comma
+        if self.suffix != "":
+            suffix=", " + self.suffix
+        else:
+            suffix=""    
+        return " ".join(filter(None, [self.first_name, self.middle_name, self.last_name])) + suffix
+
+    @full_name.setter
+    def full_name(self, full_name):
+
+        """ allow the donor information to be updated via full_name as well """
+
+        # for simplicity's sake, we make the assumption a suffix will follow a comma
+
+        # capture suffix, if present
+        suffix=""
```

Comment:
this could be simplified a bit by doing the split first (and use rsplit, maybe):

```
In [22]: "Roy Wood, Jr".rsplit(",", 1)
Out[22]: ['Roy Wood', ' Jr']

In [23]: "Roy Wood,Jr".rsplit(",", 1)
Out[23]: ['Roy Wood', 'Jr']

In [24]: "Roy Wood".rsplit(",", 1)
Out[24]: ['Roy Wood']

In [25]: "Roy H. Wood, Jr".rsplit(",", 1)
Out[25]: ['Roy H. Wood', ' Jr']
```